File: ruby-backend/app/models/room.rb
# Room model - Compatible with Laravel rooms table
require 'sequel'
require 'securerandom'

class Room < Sequel::Model
  # Use existing Laravel table structure
  set_dataset :rooms
  
  # Plugins for enhanced functionality
  plugin :validation_helpers
  plugin :timestamps, update_on_create: true
  plugin :association_dependencies
  
  # Allow setting id for UUID compatibility
  unrestrict_primary_key
  
  # UUID generation for primary key (Laravel compatibility)
  def before_create
    super
    self.id ||= SecureRandom.uuid
  end
  
  # Associations
  many_to_one :administrator, class: :User, key: :administrator_id
  many_to_one :current_track, class: :Track, key: :current_track_id
  one_to_many :tracks, class: :Track, key: :room_id
  one_to_many :participants, class: :RoomParticipant, key: :room_id
  
  # Many-to-many association with users through room_participants
  many_to_many :users, join_table: :room_participants, left_key: :room_id, right_key: :user_id
  
  # Validation rules (compatible with Laravel validation)
  def validate
    super
    validates_presence [:name, :administrator_id]
    validates_max_length 100, :name
    # Note: Foreign key validation is handled by database constraints
    # validates_foreign_key is not available in Sequel, but we can check manually if needed
  end
  
  # Class methods for validation rules (Laravel compatibility)
  def self.validation_rules
    {
      name: 'required|string|max:100',
      administrator_id: 'required|uuid|exists:users,id'
    }
  end
  
  def self.update_validation_rules
    {
      name: 'sometimes|string|max:100',
      current_track_id: 'sometimes|nullable|uuid|exists:tracks,id',
      is_playing: 'sometimes|boolean'
    }
  end
  
  # Track queue ordered by vote score (desc) then created_at (asc)
  def track_queue
    tracks_dataset.order(Sequel.desc(:vote_score), :created_at)
  end
  
  # Get next track in queue (excluding currently playing track)
  def next_track
    queue = track_queue
    
    # If there's a current track, get the next one after it
    if current_track_id
      queue = queue.exclude(id: current_track_id)
    end
    
    queue.first
  end
  
  # Get track queue as array with position information
  def track_queue_with_positions
    queue = track_queue.to_a
    queue.each_with_index.map do |track, index|
      track_data = track.to_hash
      track_data[:queue_position] = index + 1
      track_data[:is_current] = track.id == current_track_id
      track_data[:is_next] = index == 0 && track.id != current_track_id
      track_data
    end
  end
  
  # Playback position calculation (Laravel compatibility)
  def current_playback_position
    return 0 unless is_playing && playback_started_at
    
    elapsed = Time.now - playback_started_at
    
    # If there was a pause, subtract the paused duration
    if playback_paused_at && playback_paused_at > playback_started_at
      paused_duration = Time.now - playback_paused_at
      elapsed -= paused_duration
    end
    
    [0, elapsed.to_i].max
  end
  
  # Playback control methods (Laravel compatibility)
  def start_track(track)
    update(
      current_track_id: track.id,
      playback_started_at: Time.now,
      playback_paused_at: nil,
      is_playing: true
    )
  end
  
  def pause_playback
    update(
      playback_paused_at: Time.now,
      is_playing: false
    )
  end
  
  def resume_playback
    if playback_paused_at
      paused_duration = Time.now - playback_paused_at
      new_start_time = playback_started_at + paused_duration
      
      update(
        playback_started_at: new_start_time,
        playback_paused_at: nil,
        is_playing: true
      )
    end
  end
  
  def stop_playback
    update(
      current_track_id: nil,
      playback_started_at: nil,
      playback_paused_at: nil,
      is_playing: false
    )
  end
  
  def skip_to_next
    next_track_obj = next_track
    
    if next_track_obj
      start_track(next_track_obj)
    else
      stop_playback
    end
    
    next_track_obj
  end
  
  # Participant management methods (Laravel compatibility)
  def add_participant(user)
    # Check if participant already exists
    existing = participants_dataset.where(user_id: user.id).first
    return existing if existing
    
    # Create new participant with explicit ID
    RoomParticipant.create(
      id: SecureRandom.uuid,
      room_id: id,
      user_id: user.id,
      joined_at: Time.now
    )
  end
  
  def remove_participant(user)
    participants_dataset.where(user_id: user.id).delete > 0
  end
  
  def administered_by?(user)
    administrator_id == user.id
  end
  
  def has_participant?(user)
    # Administrators are always considered participants
    return true if administrator_id == user.id
    
    participants_dataset.where(user_id: user.id).count > 0
  end
  
  # Get participant count (Laravel compatibility)
  def participant_count
    participants.count
  end
  
  # Serialization for API responses (Laravel format compatibility)
  def to_hash
    {
      id: id,
      name: name,
      administrator_id: administrator_id,
      administrator: administrator&.to_hash,
      current_track_id: current_track_id,
      current_track: current_track&.to_hash,
      playback_started_at: playback_started_at&.iso8601,
      playback_paused_at: playback_paused_at&.iso8601,
      is_playing: is_playing,
      participants: participants.map(&:to_hash),
      track_queue: track_queue.map(&:to_hash),
      track_count: tracks.count,
      participant_count: participants.count,
      created_at: created_at&.iso8601,
      updated_at: updated_at&.iso8601
    }
  end
  
  def to_json(*args)
    to_hash.to_json(*args)
  end
  
  # Association dependency cleanup
  add_association_dependencies tracks: :destroy
  add_association_dependencies participants: :destroy
end

File: backend/app/Models/Room.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Concerns\HasUuids;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Database\Eloquent\Relations\BelongsToMany;
use Carbon\Carbon;

class Room extends Model
{
    use HasFactory, HasUuids;

    protected $fillable = [
        'name',
        'administrator_id',
        'current_track_id',
        'playback_started_at',
        'playback_paused_at',
        'is_playing',
    ];

    protected $casts = [
        'playback_started_at' => 'datetime',
        'playback_paused_at' => 'datetime',
        'is_playing' => 'boolean',
    ];

    /**
     * Validation rules for room creation
     */
    public static function validationRules(): array
    {
        return [
            'name' => 'required|string|max:100',
            'administrator_id' => 'required|uuid|exists:users,id',
        ];
    }

    /**
     * Validation rules for room update
     */
    public static function updateValidationRules(): array
    {
        return [
            'name' => 'sometimes|string|max:100',
            'current_track_id' => 'sometimes|nullable|uuid|exists:tracks,id',
            'is_playing' => 'sometimes|boolean',
        ];
    }

    /**
     * Room administrator
     */
    public function administrator(): BelongsTo
    {
        return $this->belongsTo(User::class, 'administrator_id');
    }

    /**
     * Currently playing track
     */
    public function currentTrack(): BelongsTo
    {
        return $this->belongsTo(Track::class, 'current_track_id');
    }

    /**
     * All tracks in this room
     */
    public function tracks(): HasMany
    {
        return $this->hasMany(Track::class);
    }

    /**
     * Room participants (through room_participants table)
     */
    public function participants(): HasMany
    {
        return $this->hasMany(RoomParticipant::class);
    }

    /**
     * Users participating in this room
     */
    public function users(): BelongsToMany
    {
        return $this->belongsToMany(User::class, 'room_participants')
                    ->withPivot('joined_at')
                    ->withTimestamps();
    }

    /**
     * Get track queue ordered by vote score (desc) then created_at (asc)
     */
    public function trackQueue()
    {
        return $this->hasMany(Track::class)
                    ->orderBy('vote_score', 'desc')
                    ->orderBy('created_at', 'asc');
    }

    /**
     * Get the next track in queue
     */
    public function getNextTrack(): ?Track
    {
        return $this->trackQueue()->first();
    }

    /**
     * Calculate current playback position in seconds
     */
    public function getCurrentPlaybackPosition(): int
    {
        if (!$this->is_playing || !$this->playback_started_at) {
            return 0;
        }

        $elapsed = Carbon::now()->diffInSeconds($this->playback_started_at);
        
        // If there was a pause, subtract the paused duration
        if ($this->playback_paused_at && $this->playback_paused_at->greaterThan($this->playback_started_at)) {
            $pausedDuration = Carbon::now()->diffInSeconds($this->playback_paused_at);
            $elapsed -= $pausedDuration;
        }

        return max(0, $elapsed);
    }

    /**
     * Start playing a track
     */
    public function startTrack(Track $track): void
    {
        $this->update([
            'current_track_id' => $track->id,
            'playback_started_at' => Carbon::now(),
            'playback_paused_at' => null,
            'is_playing' => true,
        ]);
    }

    /**
     * Pause current playback
     */
    public function pausePlayback(): void
    {
        $this->update([
            'playback_paused_at' => Carbon::now(),
            'is_playing' => false,
        ]);
    }

    /**
     * Resume paused playback
     */
    public function resumePlayback(): void
    {
        if ($this->playback_paused_at) {
            $pausedDuration = Carbon::now()->diffInSeconds($this->playback_paused_at);
            $newStartTime = $this->playback_started_at->addSeconds($pausedDuration);
            
            $this->update([
                'playback_started_at' => $newStartTime,
                'playback_paused_at' => null,
                'is_playing' => true,
            ]);
        }
    }

    /**
     * Stop playback
     */
    public function stopPlayback(): void
    {
        $this->update([
            'current_track_id' => null,
            'playback_started_at' => null,
            'playback_paused_at' => null,
            'is_playing' => false,
        ]);
    }

    /**
     * Skip to next track
     */
    public function skipToNext(): ?Track
    {
        $nextTrack = $this->getNextTrack();
        
        if ($nextTrack) {
            $this->startTrack($nextTrack);
        } else {
            $this->stopPlayback();
        }
        
        return $nextTrack;
    }

    /**
     * Add user to room
     */
    public function addParticipant(User $user): RoomParticipant
    {
        return $this->participants()->firstOrCreate([
            'user_id' => $user->id,
        ], [
            'joined_at' => Carbon::now(),
        ]);
    }

    /**
     * Remove user from room
     */
    public function removeParticipant(User $user): bool
    {
        return $this->participants()
                    ->where('user_id', $user->id)
                    ->delete() > 0;
    }

    /**
     * Check if user is administrator
     */
    public function isAdministratedBy(User $user): bool
    {
        return $this->administrator_id === $user->id;
    }

    /**
     * Check if user is participant
     */
    public function hasParticipant(User $user): bool
    {
        return $this->participants()
                    ->where('user_id', $user->id)
                    ->exists();
    }
}